source/shadow.c

```diff
@@ -245,6 +298,54 @@ static ShadowStatus_t validateMatchTopicParameters( const char * pTopic,
 
 /*-----------------------------------------------------------*/
 
+ShadowStatus_t validateAssembleTopicParameters( ShadowTopicStringType_t topicType,
+                                                const char * pThingName,
+                                                uint8_t thingNameLength,
+                                                const char * pShadowName,
+                                                uint8_t shadowNameLength,
+                                                char * pTopicBuffer,
+                                                uint16_t * pOutLength )
+{
+    ShadowStatus_t shadowStatus = SHADOW_SUCCESS;
```

Comment:
maintainability: return values should always be initialized to failure, i.e.:

```c
ShadowStatus_t shadowStatus = SHADOW_BAD_PARAMETER;

if( ( pTopicBuffer != NULL ) &&
    ( pThingName != NULL ) &&
    ( thingNameLength > 0U ) &&
    ( pShadowName != NULL ) &&
    ( shadowNameLength > 0U ) &&
    ( topicType < ShadowTopicStringTypeMaxNum ) &&
    ( pOutLength != NULL ) )
{
    if( thingNameLength <= SHADOW_THINGNAME_MAX_LENGTH )
    {
        if( shadowNameLength <= SHADOW_NAME_MAX_LENGTH )
        {
            shadowStatus = SHADOW_SUCCESS;
        }
        else
        {
            LogError( ( "Invalid shadowNameLength. Shadow name length of %u exceeds maximum allowed length %u.",
                    ( unsigned int ) shadowNameLength,
                    ( unsigned int ) SHADOW_NAME_MAX_LENGTH ) );
        }
    }
    else
    {
        LogError( ( "Invalid thingNamelength. Thing name length of %u exceeds maximum allowed length %u.",
                    ( unsigned int ) thingNameLength,
                    ( unsigned int ) SHADOW_THINGNAME_MAX_LENGTH ) );

    }
}
else
{
       LogError( ( "Invalid input parameters pTopicBuffer: %p, pThingName: %p, thingNameLength: %u,\
                    pShadowName: %p, shadowNameLength: %u, topicType: %d, pOutLength: %p.",
                    ( void * ) pTopicBuffer,
                    ( void * ) pThingName,
                    ( unsigned int ) thingNameLength,
                    ( void * ) pShadowName,
                    ( unsigned int ) shadowNameLength,
                    ( int ) topicType,
                    ( void * ) pOutLength ) );
}
return shadowStatus;
```

note that in addition to improving the maintainability of the code, defaulting to the failure case often simplifies the logic.